boa/src/builtins/object/mod.rs

```diff
@@ -104,6 +114,99 @@ impl Object {
+        let object = args.get(0).unwrap().to_object(ctx)?;
+        if let Some(key) = args.get(1) {
+            let key = key.to_property_key(ctx)?;
+            let desc = object.borrow().get_own_property(&key);
+            Self::from_property_descriptor(desc, ctx)
+        } else {
+            Ok(Value::undefined())
+        }
+    }
+
+    /// `Object.getOwnPropertyDescriptors( object )`
+    ///
+    /// Returns all own property descriptors of a given object.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://www.ecma-international.org/ecma-262/10.0/index.html#sec-object.getownpropertydescriptor
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptors
+    pub fn get_own_property_descriptors(
+        _: &Value,
+        args: &[Value],
+        ctx: &mut Context,
+    ) -> Result<Value> {
+        let object = args.get(0).unwrap_or(&Value::undefined()).to_object(ctx)?;
+        let descriptors = ctx.construct_object();
+
+        for key in object.borrow().keys() {
+            let desc = object.borrow().get_own_property(&key);
+            let descriptor = Self::from_property_descriptor(desc, ctx)?;
+
+            if !descriptor.is_undefined() {
+                descriptors
+                    .borrow_mut()
+                    .insert(key, Property::data_descriptor(descriptor, Attribute::all()));
+            }
+        }
+
+        Ok(Value::Object(descriptors))
+    }
+
+    /// https://www.ecma-international.org/ecma-262/10.0/index.html#sec-frompropertydescriptor
+    fn from_property_descriptor(desc: Property, ctx: &mut Context) -> Result<Value> {
+        let descriptor = ctx.construct_object();
+        if let Some(value) = &desc.value {
+            descriptor.borrow_mut().insert(
+                "value",
+                Property::data_descriptor(value.clone(), Attribute::all()),
+            );
+        }
+        if let Some(set) = &desc.set {
+            descriptor.borrow_mut().insert(
+                "set",
+                Property::data_descriptor(set.clone(), Attribute::all()),
+            );
+        }
+        if let Some(get) = &desc.get {
+            descriptor.borrow_mut().insert(
+                "get",
+                Property::data_descriptor(get.clone(), Attribute::all()),
+            );
+        }
+        descriptor.borrow_mut().insert(
+            "writable",
+            Property::data_descriptor(desc.writable().into(), Attribute::all()),
+        );
+        descriptor.borrow_mut().insert(
+            "enumerable",
+            Property::data_descriptor(desc.enumerable().into(), Attribute::all()),
+        );
+        descriptor.borrow_mut().insert(
+            "configurable",
+            Property::data_descriptor(desc.configurable().into(), Attribute::all()),
+        );
+        Ok(Value::Object(descriptor))
```

Comment:
Instead of adding property descriptors by hand you can use the `ObjectInitializer` (found in src/object/mod.rs) builder should make this much clearer.

Reply:
Ah thanks! I was looking for that 😄.